src/Templates/SingleProductTemplateCompatibility.php

```diff
@@ -0,0 +1,401 @@
+	 * woocommerce_before_single_product
+	 * woocommerce_before_single_product_summary
+	 * woocommerce_single_product_summary
+	 *
+	 * The last block supports the following hooks:
+	 * woocommerce_after_single_product
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	private function inject_hook_to_first_and_last_blocks( $block_content, $block ) {
+		$first_block_hook = array(
+			'before' => array(
+				'woocommerce_before_single_product'  => $this->hook_data['woocommerce_before_single_product'],
+				'woocommerce_before_single_product_summary' => $this->hook_data['woocommerce_before_single_product_summary'],
+				'woocommerce_single_product_summary' => $this->hook_data['woocommerce_single_product_summary'],
+			),
+			'after'  => array(),
+		);
+
+		$last_block_hook = array(
+			'before' => array(),
+			'after'  => array(
+				'woocommerce_after_single_product' => $this->hook_data['woocommerce_after_single_product'],
+			),
+		);
+
+		if ( isset( $block['attrs'][ self::IS_FIRST_BLOCK ] ) && isset( $block['attrs'][ self::IS_LAST_BLOCK ] ) ) {
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer(
+					$first_block_hook['before'],
+					'before'
+				),
+				$block_content,
+				$this->get_hooks_buffer(
+					array_merge(
+						$first_block_hook['after'],
+						$last_block_hook['after']
+					),
+					'after'
+				)
+			);
+		}
+
+		if ( isset( $block['attrs'][ self::IS_FIRST_BLOCK ] ) ) {
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer(
+					$first_block_hook['before'],
+					'before'
+				),
+				$block_content,
+				$this->get_hooks_buffer(
+					$first_block_hook['after'],
+					'after'
+				)
+			);
+		}
+
+		if ( isset( $block['attrs'][ self::IS_LAST_BLOCK ] ) ) {
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer( array(), 'before' ),
+				$block_content,
+				$this->get_hooks_buffer(
+					$last_block_hook['after'],
+					'after'
+				)
+			);
+		}
+	}
+
+	/**
+	 * Update the render block data to inject our custom attribute needed to
+	 * determine which is the first block of the Single Product Template.
+	 *
+	 * @param array         $parsed_block The block being rendered.
+	 * @param array         $source_block An un-modified copy of $parsed_block, as it appeared in the source content.
+	 * @param WP_Block|null $parent_block If this is a nested block, a reference to the parent block.
+	 *
+	 * @return array
+	 */
+	public function update_render_block_data( $parsed_block, $source_block, $parent_block ) {
+		return $parsed_block;
+	}
+
+	/**
+	 * Set supported hooks.
+	 */
+	protected function set_hook_data() {
+		$this->hook_data = array(
+			'woocommerce_before_single_product'         => array(
+				'block_name' => '',
+				'position'   => 'before',
+				'hooked'     => array(
+					'woocommerce_output_all_notices' => 10,
+					'woocommerce_breadcrumb'         => 20,
```

Comment:
The breadcrumb template function is hooked to `'woocommerce_before_main_content'` rather than `'woocommerce_before_single_product'`

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-hooks.php#L47

Reply:
Nice catch! Fixed via [ebe5b5e](https://github.com/woocommerce/woocommerce-blocks/pull/8442/commits/ebe5b5e825798eeb8b004fcb693f29f6e54eb71b)

Reply:
Thanks for the update! Do you think it's worth adding a compatibility layer for the breadcrumb elsewhere (or maybe just add the correct hook `'woocommerce_before_main_content'` to this array, even though it is not strict to the sing product template)?

Reply:
Mmm, why? The breadcrumb is a wrapper of the PHP code, so it already support all the hooks, no?

Reply:
> Mmm, why? The breadcrumb is a wrapper of the PHP code, so it already support all the hooks, no?

Yep! TBH I'm not sure if this is necessary or not, as I'm still not super familiar with the background decisions made for this compatibility layer. What I was wondering is if, the same way we had the `'woocommerce_breadcrumb'` hooked here under `'woocommerce_before_single_product'`, we wouldn't want to keep the same pattern and just move it to `'woocommerce_before_main_content'` instead (which is the correct one)? Maybe it's not a requirement now that it is not under `'woocommerce_before_single_product'` anymore, just wanted to double-check.

Reply:
All **the callbacks** that we put inside `hooked` are removed because we don't want that WooCommerce injects element in the DOM because now we have the blocks.

For this reason, we have this bug https://github.com/woocommerce/woocommerce-blocks/issues/8528 because we are removing the callback `woocommerce_show_product_sale_flash`. In this case, we should keep this callback (I will create a dedicated PR for this).

Does it make sense? Feel free to reach me if you have other questions!

Reply:
> All the callbacks that we put inside hooked are removed because we don't want that WooCommerce injects element in the DOM because now we have the blocks.

Thanks, but that wasn't really my question. My question is regarding the breadcrumbs: because [we were previously removing its callback](https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1113458023) from `'woocommerce_before_single_product'` (which was incorrect ), and now it was [simply removed with no replacement](https://github.com/woocommerce/woocommerce-blocks/pull/8442/commits/ebe5b5e825798eeb8b004fcb693f29f6e54eb71b) (which I believe there should be).

Reply:
Ah, gotcha! In 84d03e0f8936d789f6a26b4c00e79550e3e5d720, I added the missing hooks 💪

Reply:
Just to better illustrate, I'm talking about something along these lines:
```
			'woocommerce_before_main_content'         => array(
				'block_name' => '',
				'position'   => 'before',
				'hooked'     => array(
					'woocommerce_breadcrumb' => 20,
				),
			),
```

Reply:
> Ah, gotcha! In https://github.com/woocommerce/woocommerce-blocks/commit/84d03e0f8936d789f6a26b4c00e79550e3e5d720, I added the missing hooks 💪

Cool, thanks! I'm gonna check & re-run the tests soon.